Add support to join tables on multiple columns.

gregorco opened the pull request:
Thanks for contributing.

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](https://developercertificate.org)

## Description

Enhanced API to accept multiple names of columns to join on.  Supported for all joins: inner, leftOuter, rightOuter, and fullOuter.  Added javadoc where missing for public methods.

## Testing

Added dozens of junit tests to cover all changed API.  Ran coverage tool in eclipse to ensure testing/exercising of all changes was being done.  Reached 100% coverage for all but one method.  Goal to reach 100% coverage exposed failure of LONG types to be parse from cvs input so added missing support for that.

benmccann commented on core/src/main/java/tech/tablesaw/api/Table.java:
```diff
@@ -572,7 +572,7 @@ public void addRow(int rowIndex, Table sourceTable) {
         }
     }
-
+    
```
unnecessary whitespace

benmccann commented on core/src/main/java/tech/tablesaw/api/Table.java:
```diff
@@ -865,13 +865,14 @@ public Table countBy(CategoricalColumn<?> groupingColumn) {
+    public DataFrameJoiner join(String... columnNames) {
+        return new DataFrameJoiner(this, columnNames);
+	}
+    
```
unnecessary whitespace

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -434,42 +487,62 @@ public Table rightOuter(boolean allowDuplicateColumnNames, Table... tables) {
+            try {
+                result.addColumns(leftOuter.column(name));
+            } catch (IllegalStateException e) {
+                System.out.println("NOTE: DataFrameJoiner.rightOuter(): skipping left table's column,'"
```
We should either throw an exception here or do logging with slf4j rather than printing to std out

gregorco replied:
I don't yet see any use of slf4j in this project.  Would this be the first use?  If so, is log4j preferred over logback?  I could introduce either one but I've gotten the impression that logback is gaining over log4j in popularity.

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -26,19 +27,32 @@
-        this.column = table.categoricalColumn(column);
+        columns = new CategoricalColumn<?>[columnNames.length];
+        this.columnNames = columnNames;
+        for(int i = 0; i < this.columnNames.length; i++) {
```
add a space after `for`

benmccann replied:
Could simplify this as `for (String colName : columnNames)`

benmccann replied:
ok, no need then. I didn't see that `i` was being used anywhere else

gregorco replied:
You're right - it's not being used anywhere else in that immediate method - but later on in the processing in another method the logic relies on the same index variable coordinating the name to the column.  This seemed like the simplest/cleanest solution.  Thanks.  I'll prepare and submit a new update with fixes for the other changes soon.

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -85,142 +112,145 @@ public Table inner(Table table2, String col2Name) {
                 } else {
-                    crossProduct(result, table1Rows, table2Rows);
+                    throw new IllegalArgumentException(
+                            "Joining is supported on numeric, string, and date-like columns. Column "
```
I might just say categorical columns here because double columns are numeric, but not categorical, which makes this message a bit misleading

gregorco replied:
If message was changed to say that only categorical columns are supported then BooleanColumn should already be accepted, but trying to join on a boolean column will currently filter to this else condition; I got this condition to be exercised with a junit test that joined on a boolean column.   BooleanColumn implements CategoricalColumn though so should an if condition be added here for BooleanColumnType?

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -85,142 +112,145 @@ public Table inner(Table table2, String col2Name) {
-                } else {
-                    crossProduct(result, table1Rows, table2Rows);
+                // combine Selection's into one big AND Selection
+                if(rowBitMapOneCol != null) {
```
add a space after `if`, `?`, and `:`

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -231,135 +261,141 @@ private String newName(String table2Alias, String columnName) {
+            int ri = row.getRowNumber();
+            Selection rowBitMapMultiCol = null;
+
+            for(int i = 0; i < columns.length; i++) {
```
add a space after `for`

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -231,135 +261,141 @@ private String newName(String table2Alias, String columnName) {
-                if (index.get(value).isEmpty()) {
-                    selection.add(i);
+                // combine Selections into one big AND Selection
+                if(rowBitMapOneCol != null) {
```
space after `if`, `?`, and `:`

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -231,135 +261,141 @@ private String newName(String table2Alias, String columnName) {
-                if (index.get(value).isEmpty()) {
-                    selection.add(i);
-                }
+            if(rowBitMapMultiCol.isEmpty()) {
```
space after `if`

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -231,135 +261,141 @@ private String newName(String table2Alias, String columnName) {
+                    rowBitMapOneCol = index.get(value);
+                } else {
+                    throw new IllegalArgumentException(
+                            "Joining is supported on numeric, string, and date-like columns. Column "
```
would change message to categorical as above

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -511,21 +584,27 @@ private void withMissingLeftJoin(Table destination, Table table1) {
-            if (c < t2StartCol) {
-                for (int r2 = 0; r2 < table2.rowCount(); r2++) {
-                    destination.column(c).appendMissing();
+            if(!addedJoinColumns) {
```
space after if

benmccann commented:
Wow, this is great!

gregorco commented:
@benmccann  Thanks for the review.  I'll work on those changes.

lwhite1 commented:
Joining on Boolean columns Isnt supported.  If it was done on the Boolean
col only you would get n squared rows frequently

On Wed, Sep 5, 2018 at 11:21 AM Gregor Okorn <notifications@github.com>
wrote:

> *@gregorco* commented on this pull request.
> ------------------------------
>
> In core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java
> <https://github.com/jtablesaw/tablesaw/pull/377#discussion_r215315430>:
>
> >                  } else {
> -                    crossProduct(result, table1Rows, table2Rows);
> +                    throw new IllegalArgumentException(
> +                            "Joining is supported on numeric, string, and date-like columns. Column "
>
> If message was changed to say that only categorical columns are supported
> then BooleanColumn should already be accepted, but trying to join on a
> boolean column will currently filter to this else condition; I got this
> condition to be exercised with a junit test that joined on a boolean
> column. BooleanColumn implements CategoricalColumn though so should an if
> condition be added here for BooleanColumnType?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jtablesaw/tablesaw/pull/377#discussion_r215315430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADRXguDZRYYzS2w8T6DD1bVflEEoY05gks5uX-wWgaJpZM4WaTs4>
> .
>

gregorco commented:
Thanks for the clarification.  Since BooleanColumn implements CategoricalColumn though, should the message in that IllegalArgumentException be changed - or left as is?

benmccann commented:
Maybe change "numeric" to "integral" ?

gregorco commented:
Sounds good.  Thanks. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

gregorco commented:
@benmccann My app used to be able to join on a DoubleColumn, but now as I'm integrating recent changes into it, I'm seeing a failure that the DoubleColumn I'm trying to join on can't be cast to a CategoricalColumn.   I know I had the join working in the past.  At what point was that ability changed/restricted?  Is there any workaround to allow a join on a DoubleColumn?

lwhite1 commented:
it changed in master, i think.

(btw, I think this should be in a separate issue because it has nothing to
do with joining on multiple columns)

There was a concern that joining on double was a bit dangerous due to
rounding issues. We added joining on doubles when we didn't have any other
numeric column types. Usually people want to join on an int column.

Do you need to join on double or can your data be stored in an int column
before joining?

i'm not sure we should really prevent people from trying to join on double.
I'm also a bit uncertain that we should prevent people from joining on
booleans, because if they blow up their app it's on them. Perhaps we should
allow joining on any column type. Fewer special cases are better.


On Thu, Sep 6, 2018 at 4:07 AM Gregor Okorn <notifications@github.com>
wrote:

> @benmccann <https://github.com/benmccann> My app used to be able to join
> on a DoubleColumn, but now as I'm integrating recent changes into it, I'm
> seeing a failure that the DoubleColumn I'm trying to join on can't be cast
> to a CategoricalColumn. I know I had the join working in the past. At what
> point was that ability changed/restricted? Is there any workaround to allow
> a join on a DoubleColumn?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jtablesaw/tablesaw/pull/377#issuecomment-419004135>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADRXggZ71sWzRh4oFoaGCn_Eos6YHYjhks5uYNfBgaJpZM4WaTs4>
> .
>

gregorco commented:
@benmccann Don't know if you received notice of the reply comments I added yesterday regarding your comments for the slf4j and simplified for-loop suggestions.  I could add a logging mechanism, but am wondering if logback would be preferred over log4j.  The classic for-loop seemed to be the best solution for that situation.  Let me know what you think.  Thanks.

lwhite1 commented:
We don't use any library currently. Logging in java is a mess. In larger
projects you get multiple logging libraries all doing their own thing
because the libraries add them. I don't know that adding one is an
improvement or not.

I would prefer, I think, to write (relatively few) warnings/informational
output to standard out, and throw exceptions for real issues, but I'm open
to discussion. In the example code above, if the Illegal state causes real
issues we should let the client code deal with it.



On Thu, Sep 6, 2018 at 9:48 AM Gregor Okorn <notifications@github.com>
wrote:

> @benmccann <https://github.com/benmccann> Don't know if you received
> notice of the reply comments I added yesterday regarding your comments for
> the slf4j and simplified for-loop suggestions. I could add a logging
> mechanism, but am wondering if logback would be preferred over log4j. The
> classic for-loop seemed to be the best solution for that situation. Let me
> know what you think. Thanks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jtablesaw/tablesaw/pull/377#issuecomment-419099400>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADRXglgkN9r_l5aJPc6VWg736Y49J2FEks5uYSeegaJpZM4WaTs4>
> .
>

gregorco commented:
In the above code the exception is anticipated and not a real concern.  The informational message isn't necessary, so I can remove it altogether.

benmccann commented:
80%+ of Java libraries today use slf4j-api, which is a facade that delegates to other logging libraries. The end user can then pick any logging implementation they want to use and the log messages will be redirected to that underlying implementation. I definitely would not introduce any other library like log4j, logback, etc. because there's such widespread fragmentation between logging implementations that it will cause problems for some subset of users. But as long as you use slf4j-api it will work with any implementation that the user may choose.

The thing I don't like about writing to std out is that the end user has no control over configuring the output

lwhite1 commented:
yeah, in hindsight, i don't think we write to sout anywhere unless it's a
stray debugging statement that needs to be deleted. I'm trying to do to
many things at once to provide good feedback this morning.

Still not sure how i feel about adding slf4j as a dependency, or
philosophically how i feel about whether a library like this should be
doing logging.

On Thu, Sep 6, 2018 at 10:42 AM Ben McCann <notifications@github.com> wrote:

> 80%+ of Java libraries today use slf4j-api, which is a facade that
> delegates to other logging libraries. The end user can then pick any
> logging implementation they want to use and the log messages will be
> redirected to that underlying implementation. I definitely would not
> introduce any other library like log4j, logback, etc. because there's such
> widespread fragmentation between logging implementations that it will cause
> problems for some subset of users. But as long as you use slf4j-api it will
> work with any implementation that the user may choose.
>
> The thing I don't like about writing to std out is that the end user has
> no control over configuring the output
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jtablesaw/tablesaw/pull/377#issuecomment-419119780>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADRXgihAVaGj8Z2f_Noa08JWjqRsWAqEks5uYTRqgaJpZM4WaTs4>
> .
>

gregorco commented:
The message that I output can simply be removed.  The exception was anticipated so not a real concern or need for action.  

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

gregorco commented:
Thank you @benmccann .  I'll work on these changes.

gregorco commented:
@benmccann , @lwhite1  I submitted the changes, but the travis-ci build failed with no clear reason.  I don't see a way for me to tell it to restart the build.  Are you able to restart the build?

lwhite1 commented:
I can do that    Travis is flakey

On Thu, Sep 6, 2018 at 10:39 PM Gregor Okorn <notifications@github.com>
wrote:

> @benmccann <https://github.com/benmccann> , @lwhite1
> <https://github.com/lwhite1> I submitted the changes, but the travis-ci
> build failed with no clear reason. I don't see a way for me to tell it to
> restart the build. Are you able to restart the build?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/jtablesaw/tablesaw/pull/377#issuecomment-419302690>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADRXgjajBuQE2EAaiGyhMHPXv8dkZe5oks5uYdxzgaJpZM4WaTs4>
> .
>

gregorco commented:
Looks like it tried again by itself already?  I saw this morning that the build had completed before your comment that you'd restart it.  Did you restart it earlier, or did it do so by itself?

lwhite1 commented:
It ran by itself.  All good. I will look at your pr as soon as time
allows.  Thanks very much

On Fri, Sep 7, 2018 at 9:44 AM Gregor Okorn <notifications@github.com>
wrote:

> Looks like it tried again by itself already? I saw this morning that the
> build had completed before your comment that you'd restart it. Did you
> restart it earlier, or did it do so by itself?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/jtablesaw/tablesaw/pull/377#issuecomment-419443868>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADRXgk3OeRuv9RFNzKgwGhGPI8q_IQZrks5uYnhOgaJpZM4WaTs4>
> .
>

lwhite1 commented:
Thanks very much, @gregorco
